### DISCUSSION TITLE
How to set rigid displacement of blocks in coupling analysis of tensor mechanics and heat conduction?

### DISCUSSION THREAD

**Discussion by lucuo** [idaholab/moose#26319] [Q&A Modules: Solid mechanics]
I am now using heat conduction and tensor displacement, and there is thermal expansion of the material. As shown, I set [InclineNodisplacementBC] boundary conditions on the hypotenuse of the green block. If I want to set the rigid body displacement in the direction of the red arrow throughout the green block, how to set it.
Are there any mature code in MOOSE or which source files can be modified?

**Comment by GiudGiud** [comment]
@recuero @jiangwen84 
I think we want a normal displacement BC here

**Reply by lucuo** [reply]
You mean DirichletBC? However, the DirichletBC displacement specified at the boundary will affect the original thermal expansion strain of the block.
Or are you using other code? Can you give me a specific case? Thanks a lot

**Reply by aikubo** [reply]
Just learning MOOSE so I’ll just try to guess. 

You could try creating a pinned node in the middle of edge of the domain like [this](https://mooseframework.inl.gov/modules/tensor_mechanics/tutorials/introduction/step03a.html)

But in the boundary conditions add `value = non zero number ` for the `disp` variable in the desired axis. You could also use a function as needed with the Function systems and function BCs.

**Comment by yfun0815** [comment]
I'm also curious about this question.
I'm wondering is there any function in MOOSE which can fix a variable such as displacement or pore pressure to a constant value during nonlinear iterations. In most FEM program, this function is applicable, but I didn't find any actions in MOOSE.

**Reply by GiudGiud** [reply]
hello

> during nonlinear iterations

a separate value for each nonlinear iterations?
You can make a DirichletBC depend on the current iteration number. You could code it or you could use parsed expressions involving this postprocessor
https://mooseframework.inl.gov/source/postprocessors/NumNonlinearIterations.html

**Reply by yfun0815** [reply]
No, I just wanna set a fixed value all over the calculation. Do you mean setting a value to a boundary using DirichletBC? This is only acting on boundary, but I want to act on block.

**Reply by GiudGiud** [reply]
so you are not solving any equation on that block?

**Comment by yfun0815** [comment]
Yes, I just want to solve displacement but fix pore pressure to 0 on a block.

**Reply by GiudGiud** [reply]
You can set pore pressure using an intiial condition in that block
Then set the equation in that block to only a time derivative

In the other blocks you can have the full equation

**Reply by yfun0815** [reply]
Thanks! I got it.

**Reply by yfun0815** [reply]
> You can set pore pressure using an intiial condition in that block

Still a new question. I need to change the pore pressure to a new value at a time. But I can't set initial condition to a same variable twice in a simulation with control_tags. I can see a possible method is to restart the simulation and set initial condition again, but is there any other easier method?

**Reply by GiudGiud** [reply]
>  I can see a possible method is to restart the simulation and set initial condition again

That's a fairly well supported way of doing this. 

> but is there any other easier method?

You could write a custom predictor. Predictors are the way to modify nonlienar variables

You could also use a kernel to change the equation to hit the new value. A penalty on the difference between the variable value and the target value would suffice

**Reply by yfun0815** [reply]
> You can set pore pressure using an intiial condition in that block Then set the equation in that block to only a time derivative
> 
> In the other blocks you can have the full equation

Still an issue here. Following your method, the pore pressure can be fixed to 0 in the left and right block. However, there is no pore pressure change in the point attached to the 0 pore pressure point.
<img width="886" height="396" alt="image" src="https://github.com/user-attachments/assets/72003023-c2bd-455c-94d9-3ca1bd25eacb" />
In other words, water should flow from the middle to left and right and pore pressure should decrease, something like this. Because 0 pore pressure is equal to drained boundary.
<img width="1027" height="259" alt="image" src="https://github.com/user-attachments/assets/6ee4060c-3cae-4306-9c7f-e24d5824cf3c" />

**Reply by yfun0815** [reply]
> That's a fairly well supported way of doing this.

I tried this method. Unfortunately, if the pore pressure was set to a value before, following the same kernel settings, the pore pressure cannot be fixed to another value anymore.

**Reply by GiudGiud** [reply]
did you do a recover or a restart? A restart should allow you to apply initial conditions again. A restart is specified in the Problem block with the `restart_file_base` parameter

**Reply by yfun0815** [reply]
I confirm I'm using restart.
```
[Problem]
    kernel_coverage_check = false
    material_coverage_check = false
    boundary_restricted_node_integrity_check = false
    boundary_restricted_elem_integrity_check = false
    restart_file_base = test35_vtk/_cp/0010
    force_restart = true
    allow_initial_conditions_with_restart = true
[]

[Mesh]
    [mesh]
        type = FileMeshGenerator
        #file = mesh4.msh
        file = test35_vtk/_cp/0010
    []
[]
```
The pore pressure can be initialized again but I suspect it cannot be fixed again. My kernel setting looks like that, in which only PorousFlowMassTimeDerivative is acting on all the blocks. '${soil_blocks}' is included in the '${new_domain_blocks}'. Is that correct?
```
[Kernels]
  [gravity]
    type = Gravity
    variable = disp_z
    value = -9.8
    block = '${new_domain_blocks}'
  []

  [./stress_x]
    type = StressDivergenceTensors
    component = 0
    variable = disp_x
    block = '${new_domain_blocks}'
  [../]
  [./stress_y]
    type = StressDivergenceTensors
    component = 1
    variable = disp_y
    block = '${new_domain_blocks}'
  [../]
  [./stress_z]
    type = StressDivergenceTensors
    component = 2
    variable = disp_z
    block = '${new_domain_blocks}'
  [../]

  [effective_stress_x]
    type = PorousFlowEffectiveStressCoupling
    variable = 'disp_x'
    component = 0
    block = '${soil_blocks}'
  []

  [effective_stress_y]
    type = PorousFlowEffectiveStressCoupling
    variable = 'disp_y'
    component = 1
    block = '${soil_blocks}'
  []

  [effective_stress_z]
    type = PorousFlowEffectiveStressCoupling
    variable = 'disp_z'
    component = 2
    block = '${soil_blocks}'
  []

  [mass0]
    type = PorousFlowMassTimeDerivative
    fluid_component = 0
    variable = 'porepressure'
    block = '${new_domain_blocks}'
  []
  [flux]
    type = PorousFlowAdvectiveFlux
    use_displaced_mesh = false
    variable = porepressure
    block = '${soil_blocks}'
    gravity = '0 0 -9.8'
    fluid_component = 0
  []
  [poro_vol_exp]
    type = PorousFlowMassVolumetricExpansion
    variable = 'porepressure'
    fluid_component = 0
    block = '${soil_blocks}'
  []
[]
```

**Reply by GiudGiud** [reply]
>  but I suspect it cannot be fixed again.

What do you mean? As in the treatment you were doing on some blocks with only a time derivative?
Are you trying to apply this same treatment on `'${new_domain_blocks}'` ?
What are you seeing happen on these blocks? Does this set of blocks overlap with `soil_blocks` ?

**Reply by yfun0815** [reply]
> As in the treatment you were doing on some blocks with only a time derivative?

Yes. soil_blocks+rock_blocks=new_domain_blocks. I just want to calculate pore pressure in the soil_block and fixed it to 0 in the rock_blocks. So I did time derivative on all of the blocks, and other pore pressure relevant kernels only on soil_blocks. Is it correct?
By doing so, I can see pore pressure is fixed to 0 in rock_blocks.
<img width="1556" height="663" alt="image" src="https://github.com/user-attachments/assets/7b33f5b1-a1d8-47e3-bef2-1df8e8bae7fb" />
> Are you trying to apply this same treatment on `'${new_domain_blocks}'` ?

However, when restarting the calculation, doing the same operations and fixing to a non-zero value, the pore pressure cannot be fixed in rock_blocks.

**Reply by GiudGiud** [reply]
> So I did time derivative on all of the blocks, and other pore pressure relevant kernels only on soil_blocks. Is it correct?

If you are trying to keep things constant in rock_blocks, yes it should be correct. 

> However, when restarting the calculation, doing the same operations and fixing to a non-zero value, the pore pressure cannot be fixed in rock_blocks.

Interesting. What do you mean by "cannot be fixed" ? Do you mean it jumps to the wrong value, away from the initial condition.
Or do you mean it starts evolving through time even with only a time derivative kernel?

**Reply by yfun0815** [reply]
> If you are trying to keep things constant in rock_blocks, yes it should be correct.

Ok, the issue is, pore pressure is 0 in rock block as expected, but not 0 at the interaction point between soil and rock. See the figure. How does MOOSE treat the mutual point between mesh with different pore pressure? I expect the dashed line should be with 0 pore pressure, acting as a drained boundary, by which water can flow from soil to rock.

<img width="847" height="859" alt="image" src="https://github.com/user-attachments/assets/55aaec17-93b9-4987-82e7-0da37f3938c0" />


> Or do you mean it starts evolving through time even with only a time derivative kernel?

My bad. I mean it evolved. Seems like boundary is drained somewhere. These two are showing the results after restart. I confirm I didn't change any boundary condition and use the same kernels.

<img width="1054" height="390" alt="image" src="https://github.com/user-attachments/assets/584d4c01-2735-428f-90a8-4883b89b5707" />
<img width="1040" height="338" alt="image" src="https://github.com/user-attachments/assets/ddd2cbaa-7891-450b-8ec3-1921f9c47c79" />

**Reply by GiudGiud** [reply]
You are using linear lagrange to represent the pore pressure I think. Because it is a continuous finite element family, the pore pressure value has to transition sharply over that node. 

Did you flip step 1 and step 2 here? It seems step 2 has the 0 pore pressure in the rock blocks

**Reply by GiudGiud** [reply]
> How does MOOSE treat the mutual point between mesh with different pore pressure?

MOOSE solves the nonlinear problem as posed. You have a nodal variable with a degree of freedom at every node, and you integrate an equation over quadrature points over each domain.

>  I expect the dashed line should be with 0 pore pressure, acting as a drained boundary, by which water can flow from soil to rock.

Would that really be the result of a single line of nodes with 0 pore pressure? Shouldnt you need elements to make a channel in which water can flow? I don't know.

**Reply by yfun0815** [reply]
> Did you flip step 1 and step 2 here? It seems step 2 has the 0 pore pressure in the rock blocks

No, I didn't. Rock has a super low pore pressure but not 0 in step 2. Both step 1 and step 2 are from restarting.


> Would that really be the result of a single line of nodes with 0 pore pressure? Shouldnt you need elements to make a channel in which water can flow? I don't know.

I see rock has a very large permeability, and linking to air, so the pore pressure should be nearly 0. 

> Because it is a continuous finite element family, the pore pressure value has to transition sharply over that node.

I think I did set a 0 pore pressure to rock. However, it is overwritten by a non-zero pore pressure by soil at this point.

**Reply by GiudGiud** [reply]
> I think I did set a 0 pore pressure to rock. However, it is overwritten by a non-zero pore pressure by soil at this point.

Try flipping the ICs in the input? We should be able to get one to execute after the other.

**Reply by yfun0815** [reply]
> Try flipping the ICs in the input? We should be able to get one to execute after the other.

No difference. I don't think there is difference in executing order.
Now my method is to apply a 0 pore pressure BC to this dashed line. Just like helping the kernel to apply 0 pore pressure. It looks like fitting my expectation, but it's a little bit silly by doing so. 
<img width="740" height="264" alt="image" src="https://github.com/user-attachments/assets/623d8ec9-0d9c-414f-a593-31f4e856cbad" />

**Reply by GiudGiud** [reply]
> No difference. I don't think there is difference in executing order.

you are right. The IC projection algorithm will actually use the values from the two ICs and blend them. To get continuity to work as expected. 
We could adapt it notably by leveraging the boundary parameter
